java-frontend/src/main/java/org/sonar/java/ast/visitors/FileLinesVisitor.java

```diff
@@ -19,63 +19,191 @@
+import static org.sonar.plugins.java.api.tree.Tree.Kind.STRING_LITERAL;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.TOKEN;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.TRY_STATEMENT;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.VARIABLE;
+import static org.sonar.plugins.java.api.tree.Tree.Kind.WHILE_STATEMENT;
+
 /**
  * Saves information about lines directly into Sonar by using {@link FileLinesContext}.
  */
 public class FileLinesVisitor extends SubscriptionVisitor {
 
   private final SonarComponents sonarComponents;
-  private final Charset charset;
-  private final Set<Integer> linesOfCode = Sets.newHashSet();
-  private final Set<Integer> linesOfComments = Sets.newHashSet();
+  private final Set<Integer> linesOfCode = new HashSet<>();
+  private final Set<Integer> linesOfComments = new HashSet<>();
+  private final Set<Integer> executableLines = new HashSet<>();
 
-  public FileLinesVisitor(SonarComponents sonarComponents, Charset charset) {
+  public FileLinesVisitor(SonarComponents sonarComponents) {
     this.sonarComponents = sonarComponents;
-    this.charset = charset;
   }
 
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return ImmutableList.of(Tree.Kind.TOKEN);
+    if(sonarComponents.isSQGreaterThan62()) {
+      return ImmutableList.of(TOKEN,
+        METHOD, CONSTRUCTOR,
+        INITIALIZER, STATIC_INITIALIZER,
+        VARIABLE,
+        FOR_EACH_STATEMENT, FOR_STATEMENT, WHILE_STATEMENT, DO_STATEMENT,
+        LAMBDA_EXPRESSION);
+    }
+    return ImmutableList.of(TOKEN);
   }
 
   @Override
   public void scanFile(JavaFileScannerContext context) {
     super.scanFile(context);
-
-    FileLinesContext fileLinesContext = sonarComponents.fileLinesContextFor(context.getFile());
-    // TODO minimize access to files, another one in LinesVisitor
-    int fileLength;
-    try {
-      fileLength = Files.readLines(context.getFile(), charset).size();
-    } catch (IOException e) {
-      throw Throwables.propagate(e);
-    }
+    File currentFile = context.getFile();
+    FileLinesContext fileLinesContext = sonarComponents.fileLinesContextFor(currentFile);
+    int fileLength = sonarComponents.fileLength(currentFile);
     for (int line = 1; line <= fileLength; line++) {
       fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, linesOfCode.contains(line) ? 1 : 0);
       fileLinesContext.setIntValue(CoreMetrics.COMMENT_LINES_DATA_KEY, line, linesOfComments.contains(line) ? 1 : 0);
+      if(sonarComponents.isSQGreaterThan62()) {
+        fileLinesContext.setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, line, executableLines.contains(line) ? 1 : 0);
+      }
     }
     fileLinesContext.save();
 
     linesOfCode.clear();
     linesOfComments.clear();
+    executableLines.clear();
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    List<? extends Tree> trees = Collections.emptyList();
+    switch (tree.kind()) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Complete cases by adding the missing enum constants or add a default case to this switch. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)